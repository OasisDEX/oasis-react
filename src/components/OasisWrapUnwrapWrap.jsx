import React, { PureComponent } from "react";
import { PropTypes } from "prop-types";
// import ImmutablePropTypes from 'react-immutable-proptypes';

import OasisTokenWrapFormWrapper from "../containers/OasisTokenWrapForm";
import OasisTokenBalanceWrapper from "../containers/OasisTokenBalance";
import { TOKEN_ETHER } from "../constants";
import OasisWidgetFrame from "../containers/OasisWidgetFrame";
import OasisEtherBalanceWrapper from "../containers/OasisEtherBalance";
import {
  WRAP_STATUS_VIEW_TYPE_WRAP,
  WrapUnwrapStatusWrapper
} from "../containers/WrapUnwrapStatus";
import OasisTokenBalanceSummary from "../containers/OasisTokenBalanceSummary";
import styles from "./OasisWrapUnwrapWrap.scss";
import CSSModules from "react-css-modules";
import {
  TX_WRAP_ETHER,
  TX_WRAP_TOKEN_WRAPPER
} from "../store/reducers/transactions";

const propTypes = PropTypes && {
  activeUnwrappedToken: PropTypes.string,
  onSubmit: PropTypes.func.isRequired,
  activeUnwrappedTokenBalance: PropTypes.object,
  transactionState: PropTypes.object,
  onFormChange: PropTypes.func.isRequired,
  disableForm: PropTypes.bool,
  hidden: PropTypes.bool.isRequired,
  txType: PropTypes.oneOf([TX_WRAP_ETHER, TX_WRAP_TOKEN_WRAPPER])
};
const defaultProps = {};

class OasisWrapUnwrapWrap extends PureComponent {
  constructor(props) {
    super(props);
    this.onFormChange = this.onFormChange.bind(this);
  }

  getBalance() {
    const { activeUnwrappedToken } = this.props;
    if (activeUnwrappedToken === TOKEN_ETHER) {
      return <OasisEtherBalanceWrapper decimalPlaces={5} fromWei />;
    } else {
      return (
        <OasisTokenBalanceWrapper
          decimalPlaces={5}
          tokenName={activeUnwrappedToken}
        />
      );
    }
  }

  shouldDisableForm() {
    const { disableForm } = this.props;
    return disableForm;
  }

  onFormChange() {
    const { onFormChange, anyTouched } = this.props;
    onFormChange && onFormChange(anyTouched);
  }

  render() {
    const {
      activeUnwrappedToken,
      hidden,
      txType,
      form,
      transactionState
    } = this.props;
    return (
      <OasisWidgetFrame hidden={hidden} heading={"Wrap"} spaceForContent={true}>
        <OasisTokenBalanceSummary summary="Wallet">
          {this.getBalance()}
        </OasisTokenBalanceSummary>
        <OasisTokenWrapFormWrapper
          form={form}
          txType={txType}
          onFormChange={this.onFormChange}
          transactionState={transactionState}
          activeUnwrappedToken={activeUnwrappedToken}
          onSubmit={this.props.onSubmit}
          disabled={this.shouldDisableForm()}
        />
        <WrapUnwrapStatusWrapper type={WRAP_STATUS_VIEW_TYPE_WRAP} />
      </OasisWidgetFrame>
    );
  }
}

OasisWrapUnwrapWrap.displayName = "OasisWrapUnwrapWrap";
OasisWrapUnwrapWrap.propTypes = propTypes;
OasisWrapUnwrapWrap.defaultProps = defaultProps;
export default CSSModules(OasisWrapUnwrapWrap, styles);
