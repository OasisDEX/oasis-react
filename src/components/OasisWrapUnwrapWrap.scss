.heading.heading {
  margin-bottom: 0;
}
