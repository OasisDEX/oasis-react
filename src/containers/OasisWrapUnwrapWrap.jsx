import React, { PureComponent } from 'react';
import { PropTypes } from 'prop-types';
// import ImmutablePropTypes from 'react-immutable-proptypes';

import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import wrapUnwrap from '../store/selectors/wrapUnwrap';
import OasisWrapUnwrapWrap from '../components/OasisWrapUnwrapWrap';
import wrapUnwrapReducer from '../store/reducers/wrapUnwrap';
import {
  TX_STATUS_AWAITING_CONFIRMATION,
  TX_STATUS_AWAITING_USER_ACCEPTANCE,
  TX_STATUS_CONFIRMED, TX_STATUS_REJECTED,
} from '../store/reducers/transactions';
import accounts from '../store/selectors/accounts';

const propTypes = PropTypes && {
  actions: PropTypes.object.isRequired
};

export class OasisWrapUnwrapWrapWrapper extends PureComponent {


  constructor(props) {
    super(props);
    this.state = {};
    this.makeWrap = this.makeWrap.bind(this);
  }

  makeWrap() {
    this.setState(
      {
        disableTransferButton: true,
        txStatus: false,
        txStartTimestamp: undefined
      },
      () =>
        this.props.actions.wrapToken({
          onStart: this.onTransactionStart.bind(this),
          onCancelCleanup: this.onTransactionCancelledByUser.bind(this),
          onPending: this.onTransactionPending.bind(this),
          onCompleted: this.onTransactionCompleted.bind(this),
          onRejected: this.onTransactionRejected.bind(this)
        })
    );
  }

  onTransactionStart(txStartTimestamp, txStartMeta) {
    this.setState({
      txStatus: TX_STATUS_AWAITING_USER_ACCEPTANCE,
      disableForm: true,
      lockCancelButton: true,
      txStartMeta
    });
  }

  onTransactionCancelledByUser() {
    this.setState({
      txStatus: undefined,
      disableForm: false,
      lockCancelButton: false
    });
    this.props.actions.resetActiveWrapForm();
  }
  onTransactionPending({ txStartTimestamp }) {
    this.setState({
      txStatus: TX_STATUS_AWAITING_CONFIRMATION,
      txStartTimestamp
    });
  }

  onTransactionCompleted(to, hasNextTransaction) {
    this.setState({
      txStatus: TX_STATUS_CONFIRMED
    });

    if (hasNextTransaction) {
      setTimeout(
        () => {
          this.props.actions.resetActiveWrapForm();
          this.setState({
            txStatus: undefined,
            txStartTimestamp: undefined,
            // disableForm: false
          });
        }, 3000
      );
    } else {
      setTimeout(
        () => {
          this.props.actions.resetActiveWrapForm();
          this.setState({
            txStatus: undefined,
            txStartTimestamp: undefined,
            disableForm: false
          });
        }, 3000
      )

    }
  }

  onTransactionRejected({ txHash }) {
    this.setState({
      txStatus: TX_STATUS_REJECTED,
      txHash,
      disableForm: false,
    });
  }


  render() {
    const { activeUnwrappedToken, activeUnwrappedTokenBalance } = this.props;
    const { txStatus, txStartTimestamp, txStartMeta } = this.state;
    return (
      <OasisWrapUnwrapWrap
        transactionState={{ txStatus, txStartTimestamp, txStartMeta }}
        onSubmit={this.makeWrap}
        activeUnwrappedToken={activeUnwrappedToken}
        activeUnwrappedTokenBalance={activeUnwrappedTokenBalance}
      />
    );
  }
}

export function mapStateToProps(state) {
  return {
    defaultAccount: accounts.defaultAccount(state),
    activeUnwrappedToken: wrapUnwrap.activeUnwrappedToken(state),
    activeUnwrappedTokenBalance: wrapUnwrap.activeUnwrappedTokenBalance(state)
  };
}
export function mapDispatchToProps(dispatch) {
  const actions = {
    wrapToken: wrapUnwrapReducer.actions.wrapTokenEpic,
    resetActiveWrapForm: wrapUnwrapReducer.actions.resetActiveWrapForm,
  };
  return { actions: bindActionCreators(actions, dispatch) };
}

OasisWrapUnwrapWrapWrapper.propTypes = propTypes;
OasisWrapUnwrapWrapWrapper.displayName = 'OasisWrapUnwrapWrap';
export default connect(mapStateToProps, mapDispatchToProps)(OasisWrapUnwrapWrapWrapper);
