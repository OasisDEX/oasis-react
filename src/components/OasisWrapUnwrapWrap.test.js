/* global shallow describe it expect */
/* eslint-disable import/first,no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import OasisWrapUnwrapWrap from './OasisWrapUnwrapWrap';
import BigNumber from 'bignumber.js';

describe('(Component) OasisWrapUnwrapWrap', () => {
  it('should render', () => {
    const props = {
      onSubmit: jest.fn,
      onFormChange: jest.fn,
      activeUnwrappedTokenBalance: new BigNumber(100),
      transactionState: { txStatus: true },
      hidden: false
    };
    const wrapper = shallow(
      <OasisWrapUnwrapWrap {...props}/>,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
