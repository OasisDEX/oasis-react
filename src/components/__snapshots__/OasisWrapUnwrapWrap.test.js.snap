// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) OasisWrapUnwrapWrap should render 1`] = `
<OasisWidgetFrame
  className=""
  heading="Wrap"
  hidden={false}
  spaceForContent={true}
>
  <OasisTokenBalance
    className=""
    summary="Wallet"
  >
    <Connect(OasisTokenBalance)
      decimalPlaces={5}
    />
  </OasisTokenBalance>
  <Connect(ReduxForm)
    onFormChange={[Function]}
    onSubmit={[Function]}
    transactionState={
      Object {
        "txStatus": true,
      }
    }
  />
  <WrapUnwrapStatus
    type="wrap"
  />
</OasisWidgetFrame>
`;
